In [1]:
import pandas as pd
import numpy as np

In [2]:
jan=pd.read_parquet("C:\Gaurav Personal\yellow_tripdata_2026-01.parquet")
feb=pd.read_parquet("C:\Gaurav Personal\yellow_tripdata_2026-02.parquet")
mar=pd.read_parquet("C:\Gaurav Personal\yellow_tripdata_2026-03.parquet")
apr=pd.read_parquet("C:\Gaurav Personal\yellow_tripdata_2026-04.parquet")
df=pd.concat([jan,feb,mar,apr],ignore_index=True)


In [4]:
df.shape

(14908446, 20)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14908446 entries, 0 to 14908445
Data columns (total 20 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee         

In [19]:
df.duplicated().sum()

np.int64(0)

In [20]:
df.isnull().sum()

VendorID                       0
tpep_pickup_datetime           0
tpep_dropoff_datetime          0
passenger_count          3856909
trip_distance                  0
RatecodeID               3856909
store_and_fwd_flag       3856909
PULocationID                   0
DOLocationID                   0
payment_type                   0
fare_amount                    0
extra                          0
mta_tax                        0
tip_amount                     0
tolls_amount                   0
improvement_surcharge          0
total_amount                   0
congestion_surcharge     3856909
Airport_fee              3856909
cbd_congestion_fee             0
dtype: int64

In [21]:
(df['trip_distance'] <= 0).sum()

np.int64(463713)

In [22]:
(df['fare_amount'] <= 0).sum()

np.int64(111594)

In [23]:
(df['total_amount'] <= 0).sum()

np.int64(105485)

In [24]:
(df['passenger_count'] <= 0).sum()

np.int64(52330)

In [25]:
df.describe()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
count,1.490845e+07,14908446,14908446,1.105154e+07,1.490845e+07,1.105154e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.490845e+07,1.105154e+07,1.105154e+07,1.490845e+07
mean,1.876363e+00,2026-03-03 01:02:38.999462,2026-03-03 01:20:13.606354,1.240862e+00,5.953935e+00,5.499357e+00,1.616330e+02,1.610744e+02,8.659613e-01,2.120696e+01,1.083186e+00,4.872148e-01,2.780540e+00,5.124209e-01,9.513120e-01,2.984649e+01,2.178814e+00,1.535593e-01,5.263720e-01
min,1.000000e+00,2001-01-01 09:23:58,2001-01-01 16:09:38,0.000000e+00,0.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,0.000000e+00,-2.555200e+03,-7.500000e+00,-5.000000e-01,-2.220000e+02,-1.117900e+02,-1.000000e+00,-2.560200e+03,-2.500000e+00,-2.000000e+00,-7.500000e-01
25%,2.000000e+00,2026-02-01 00:15:22,2026-02-01 00:31:04,1.000000e+00,1.000000e+00,1.000000e+00,1.140000e+02,1.070000e+02,0.000000e+00,1.000000e+01,0.000000e+00,5.000000e-01,0.000000e+00,0.000000e+00,1.000000e+00,1.728000e+01,2.500000e+00,0.000000e+00,0.000000e+00
50%,2.000000e+00,2026-03-03 19:47:48,2026-03-03 20:01:23,1.000000e+00,1.820000e+00,1.000000e+00,1.610000e+02,1.620000e+02,1.000000e+00,1.560000e+01,0.000000e+00,5.000000e-01,2.000000e+00,0.000000e+00,1.000000e+00,2.355000e+01,2.500000e+00,0.000000e+00,7.500000e-01
75%,2.000000e+00,2026-04-01 21:33:36,2026-04-01 21:49:21,1.000000e+00,3.720000e+00,1.000000e+00,2.330000e+02,2.340000e+02,1.000000e+00,2.680000e+01,2.500000e+00,5.000000e-01,3.930000e+00,0.000000e+00,1.000000e+00,3.463000e+01,2.500000e+00,0.000000e+00,7.500000e-01
max,7.000000e+00,2026-05-01 00:01:28,2026-05-02 12:06:09,9.000000e+00,3.285222e+05,9.900000e+01,2.650000e+02,2.650000e+02,4.000000e+00,2.555200e+03,2.071000e+01,1.050000e+01,7.660000e+02,1.400000e+03,1.000000e+00,2.560200e+03,2.500000e+00,2.700000e+01,7.500000e-01
std,7.179321e-01,NaN,NaN,6.470661e-01,5.904774e+02,2.029540e+01,6.681844e+01,7.083842e+01,6.579151e-01,1.862553e+01,1.733415e+00,9.769054e-02,3.910553e+00,2.176706e+00,2.442837e-01,2.230817e+01,8.893965e-01,5.490895e-01,3.503219e-01


In [26]:
df[df['trip_distance'] <= 0][['trip_distance','fare_amount','total_amount']].head()

,trip_distance,fare_amount,total_amount
118,0.0,3.0,8.00
329,0.0,18.4,24.15
330,0.0,12.8,23.15
360,0.0,50.0,54.25
634,0.0,50.0,54.25


In [27]:
df['trip_distance'].describe()

count    1.490845e+07
mean     5.953935e+00
std      5.904774e+02
min      0.000000e+00
25%      1.000000e+00
50%      1.820000e+00
75%      3.720000e+00
max      3.285222e+05
Name: trip_distance, dtype: float64

In [28]:
df[df['trip_distance'] > 100][
    ['trip_distance','fare_amount','total_amount']
].sort_values('trip_distance', ascending=False).head(20)

,trip_distance,fare_amount,total_amount
6826279,328522.20,31.11,35.86
7058668,298292.14,74.10,75.60
10403469,288381.68,21.39,31.37
14906955,281576.08,35.90,39.95
3524923,269097.48,12.19,16.94
14224153,241102.51,17.25,22.00
6822975,240830.74,15.07,20.57
6187804,237905.70,44.19,48.94
3632367,237567.19,72.99,81.95
3285750,236216.69,13.58,21.33


In [29]:
(df['trip_distance'] > 100).sum()

np.int64(631)

Identified 631 records (0.004% of the dataset) with implausible trip distances exceeding 100 miles. These records were treated as outliers and excluded from downstream analysis to improve data quality.

In [34]:
df_clean = df[df['trip_distance'] <= 100]
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.20,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0.0,0.90,1.0,N,163,162,2,7.90,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0.0,1.40,1.0,N,43,237,1,10.70,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.70,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0.0,2.16,1.0,N,88,144,1,13.50,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14908441,2,2026-04-30 23:02:00,2026-04-30 23:08:00,NaN,1.01,NaN,None,239,142,0,16.16,0.00,0.5,2.00,0.0,1.0,22.16,NaN,NaN,0.00
14908442,2,2026-04-30 23:08:31,2026-04-30 23:33:16,NaN,5.36,NaN,None,211,239,0,42.56,0.00,0.5,0.00,0.0,1.0,47.31,NaN,NaN,0.75
14908443,2,2026-04-30 23:10:19,2026-04-30 23:31:11,NaN,3.07,NaN,None,256,61,0,20.60,0.00,0.5,0.00,0.0,1.0,22.10,NaN,NaN,0.00
14908444,2,2026-04-30 23:46:47,2026-05-01 00:07:22,NaN,3.57,NaN,None,25,225,0,28.05,0.00,0.5,0.00,0.0,1.0,35.16,NaN,NaN,0.00


In [35]:
df_clean.to_parquet(
    "nyc_taxi_cleaned_v1.parquet",
    index=False
)

In [36]:
import pandas as pd

test = pd.read_parquet("nyc_taxi_cleaned_v1.parquet")
print(test.shape)

(14907815, 20)


In [5]:
df[['fare_amount']].describe()

,fare_amount
count,1.490845e+07
mean,2.120696e+01
std,1.862553e+01
min,-2.555200e+03
25%,1.000000e+01
50%,1.560000e+01
75%,2.680000e+01
max,2.555200e+03


In [6]:
df[df['fare_amount'] < 0][
    ['trip_distance','fare_amount','total_amount']
].head(20)

,trip_distance,fare_amount,total_amount
74,0.63,-5.1,-10.10
248,1.92,-22.6,-28.35
368,0.02,-3.0,-8.75
428,0.03,-3.0,-8.75
649,3.12,-22.6,-28.35
871,1.03,-8.6,-13.60
980,0.01,-70.0,-71.50
1046,0.14,-3.7,-9.45
1076,3.63,-17.7,-22.70
1106,0.01,-3.0,-8.00


In [7]:
(df['fare_amount'] < 0).sum()

np.int64(101236)

In [10]:
df_clean = df[df['fare_amount'] >= 0].copy()
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,7.20,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0.0,0.90,1.0,N,163,162,2,7.90,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0.0,1.40,1.0,N,43,237,1,10.70,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,38.70,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0.0,2.16,1.0,N,88,144,1,13.50,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14908441,2,2026-04-30 23:02:00,2026-04-30 23:08:00,NaN,1.01,NaN,None,239,142,0,16.16,0.00,0.5,2.00,0.0,1.0,22.16,NaN,NaN,0.00
14908442,2,2026-04-30 23:08:31,2026-04-30 23:33:16,NaN,5.36,NaN,None,211,239,0,42.56,0.00,0.5,0.00,0.0,1.0,47.31,NaN,NaN,0.75
14908443,2,2026-04-30 23:10:19,2026-04-30 23:31:11,NaN,3.07,NaN,None,256,61,0,20.60,0.00,0.5,0.00,0.0,1.0,22.10,NaN,NaN,0.00
14908444,2,2026-04-30 23:46:47,2026-05-01 00:07:22,NaN,3.57,NaN,None,25,225,0,28.05,0.00,0.5,0.00,0.0,1.0,35.16,NaN,NaN,0.00


In [11]:
df_clean = df[
    (df['trip_distance'] <= 100) &
    (df['fare_amount'] >= 0)
].copy()

In [12]:
print(df_clean.shape)

(14806594, 20)


In [13]:
(df_clean['fare_amount'] < 0).sum()

np.int64(0)

In [14]:
print(df_clean.shape)

(14806594, 20)


Then you removed rows with negative fares. Some of those negative-fare rows were already among the 631 outliers, so the total removed is 101,852, not 631 + 111,594

In [15]:
df_clean.to_parquet("nyc_taxi_cleaned_v1.parquet", index=False)

In [16]:
df_clean['trip_duration_minutes'] = (
    df_clean['tpep_dropoff_datetime'] - df_clean['tpep_pickup_datetime']
).dt.total_seconds() / 60

df_clean['trip_duration_minutes'].describe()

count    1.480659e+07
mean     1.759032e+01
std      2.497566e+01
min     -1.866667e+02
25%      8.216667e+00
50%      1.366667e+01
75%      2.176667e+01
max      9.983933e+03
Name: trip_duration_minutes, dtype: float64

In [17]:
(df_clean['trip_duration_minutes'] <= 0).sum()

np.int64(184306)

In [18]:
df_clean = df_clean[df_clean['trip_duration_minutes'] > 0].copy()

In [19]:
(df_clean['trip_duration_minutes'] <= 0).sum()

np.int64(0)

In [20]:
df_clean[df_clean['trip_duration_minutes'] > 300][
    ['trip_duration_minutes', 'trip_distance', 'fare_amount']
].sort_values('trip_duration_minutes', ascending=False).head(20)

,trip_duration_minutes,trip_distance,fare_amount
11561810,9983.933333,0.00,3.70
5420879,9265.150000,0.00,4.40
8805594,7620.600000,0.00,3.00
1560244,7507.900000,0.00,70.00
7524635,7482.066667,0.07,21.20
9606280,7435.666667,2.47,45.70
9288806,6746.933333,18.50,70.20
10768674,6477.333333,3.80,13.25
6895151,6335.866667,2.60,23.21
572461,5954.233333,0.00,3.70


In [21]:
(df_clean['trip_duration_minutes'] > 120).sum()

np.int64(17402)

In [22]:
(df_clean['trip_duration_minutes'] > 180).sum()

np.int64(5634)

In [23]:
(df_clean['trip_duration_minutes'] > 300).sum()

np.int64(4437)

In [25]:
df_clean['trip_duration_minutes'].describe(
    percentiles=[0.90, 0.95, 0.99, 0.995, 0.999]
)

count    1.462229e+07
mean     1.781205e+01
std      2.505382e+01
min      1.666667e-02
50%      1.381667e+01
90%      3.336667e+01
95%      4.445000e+01
99%      7.411667e+01
99.5%    8.748333e+01
99.9%    1.247000e+02
max      9.983933e+03
Name: trip_duration_minutes, dtype: float64

In [26]:
(df_clean['trip_duration_minutes'] > 125).sum()

np.int64(14457)

In [27]:
(df_clean['trip_duration_minutes'] > 180).sum()

np.int64(5634)

In [28]:
(df_clean['trip_duration_minutes'] > 300).sum()

np.int64(4437)

In [29]:
df_clean = df_clean[df_clean['trip_duration_minutes'] <= 180].copy()

In [30]:
df_clean['trip_duration_minutes'].max()

179.96666666666667

In [31]:
print(df_clean.shape)


(14616654, 21)


In [32]:
df_clean.to_parquet(
    "nyc_taxi_cleaned_v1.parquet",
    index=False
)

In [33]:
(df_clean['tip_amount'] < 0).sum()

np.int64(0)

In [34]:
(df_clean['tolls_amount'] < 0).sum()

np.int64(0)

In [35]:
(df_clean['extra'] < 0).sum()

np.int64(229)

In [36]:
df_clean['payment_type'].value_counts()

payment_type
1    9380384
0    3856149
2    1256668
4      82646
3      40807
Name: count, dtype: int64

In [37]:
df_clean[df_clean['extra'] < 0][
    ['extra', 'fare_amount', 'total_amount', 'payment_type']
].head(20)

,extra,fare_amount,total_amount,payment_type
2690908,-0.75,82.77,97.90,0
2691250,-0.75,80.00,95.13,0
2692365,-0.75,92.39,107.52,0
2694745,-0.75,108.76,123.89,0
2751996,-0.75,80.00,97.13,0
2761882,-0.75,180.48,195.61,0
2796195,-0.75,90.69,107.82,0
2817819,-0.75,80.00,95.13,0
2832654,-0.75,80.00,97.13,0
2842433,-0.75,85.93,102.47,0


In [38]:
df_clean['extra'].value_counts().sort_index()

extra
-4.75           1
-2.25           1
-0.75         227
 0.00     8530101
 0.22           1
           ...   
 14.50          1
 15.25          9
 16.71          1
 17.46          2
 20.71          1
Name: count, Length: 65, dtype: int64

In [39]:
df_clean[df_clean['extra'] == -0.75].shape

(227, 21)

In [41]:
df_clean.to_parquet(
    "nyc_taxi_cleaned_final.parquet",
    index=False
)
df_clean

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,RatecodeID,store_and_fwd_flag,PULocationID,DOLocationID,payment_type,...,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,Airport_fee,cbd_congestion_fee,trip_duration_minutes
0,2,2026-01-01 00:54:04,2026-01-01 00:59:37,1.0,0.97,1.0,N,239,238,1,...,1.00,0.5,3.66,0.0,1.0,15.86,2.5,0.0,0.00,5.550000
1,1,2026-01-01 00:34:04,2026-01-01 00:39:47,0.0,0.90,1.0,N,163,162,2,...,4.25,0.5,0.00,0.0,1.0,13.65,2.5,0.0,0.75,5.716667
2,1,2026-01-01 00:57:06,2026-01-01 01:05:59,0.0,1.40,1.0,N,43,237,1,...,4.25,0.5,2.50,0.0,1.0,18.95,2.5,0.0,0.75,8.883333
3,2,2026-01-01 00:15:22,2026-01-01 00:58:10,4.0,5.58,1.0,N,142,209,1,...,1.00,0.5,11.11,0.0,1.0,55.56,2.5,0.0,0.75,42.800000
4,2,2026-01-01 00:27:13,2026-01-01 00:40:43,0.0,2.16,1.0,N,88,144,1,...,1.00,0.5,3.85,0.0,1.0,23.10,2.5,0.0,0.75,13.500000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
14908441,2,2026-04-30 23:02:00,2026-04-30 23:08:00,NaN,1.01,NaN,None,239,142,0,...,0.00,0.5,2.00,0.0,1.0,22.16,NaN,NaN,0.00,6.000000
14908442,2,2026-04-30 23:08:31,2026-04-30 23:33:16,NaN,5.36,NaN,None,211,239,0,...,0.00,0.5,0.00,0.0,1.0,47.31,NaN,NaN,0.75,24.750000
14908443,2,2026-04-30 23:10:19,2026-04-30 23:31:11,NaN,3.07,NaN,None,256,61,0,...,0.00,0.5,0.00,0.0,1.0,22.10,NaN,NaN,0.00,20.866667
14908444,2,2026-04-30 23:46:47,2026-05-01 00:07:22,NaN,3.57,NaN,None,25,225,0,...,0.00,0.5,0.00,0.0,1.0,35.16,NaN,NaN,0.00,20.583333


In [3]:
import pandas as pd

test = pd.read_parquet("nyc_taxi_cleaned_final.parquet")
print(test.shape)

(14616654, 21)


In [4]:
import pandas as pd

numeric_cols = [
    "VendorID",
    "passenger_count",
    "trip_distance",
    "RatecodeID",
    "PULocationID",
    "DOLocationID",
    "payment_type",
    "fare_amount",
    "extra",
    "mta_tax",
    "tip_amount",
    "tolls_amount",
    "improvement_surcharge",
    "total_amount",
    "congestion_surcharge",
    "Airport_fee",
    "cbd_congestion_fee",
    "trip_duration_minutes"
]

for col in numeric_cols:
    test[col] = (
        pd.to_numeric(test[col], errors="coerce")
        .fillna(0)
    )

In [5]:
import pandas as pd

# Numeric columns
numeric_cols = [
    "VendorID",
    "passenger_count",
    "trip_distance",
    "RatecodeID",
    "PULocationID",
    "DOLocationID",
    "payment_type",
    "fare_amount",
    "extra",
    "mta_tax",
    "tip_amount",
    "tolls_amount",
    "improvement_surcharge",
    "total_amount",
    "congestion_surcharge",
    "Airport_fee",
    "cbd_congestion_fee",
    "trip_duration_minutes"
]

# String columns
string_cols = [
    "store_and_fwd_flag",
    "tpep_pickup_datetime",
    "tpep_dropoff_datetime"
]

# Clean numeric columns
for col in numeric_cols:
    test[col] = pd.to_numeric(test[col], errors="coerce").fillna(0)

# Clean string columns
for col in string_cols:
    test[col] = (
        test[col]
        .fillna("")      # Replace NULL with empty string
        .astype(str)     # Convert to string
        .str.strip()     # Remove leading/trailing spaces
    )

In [43]:
import os

os.getcwd()

'C:\\Users\\gaura\\Python'

In [6]:
test.to_csv(
    r"C:\ProgramData\MySQL\MySQL Server 8.0\Uploads\cleaned_data_fixed.csv",
    index=False
)

In [46]:
import pandas as pd

test = pd.read_parquet("nyc_taxi_cleaned_final.parquet")
print(test.shape)

(14616654, 21)


In [7]:
print(test.isnull().sum())

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
RatecodeID               0
store_and_fwd_flag       0
PULocationID             0
DOLocationID             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
Airport_fee              0
cbd_congestion_fee       0
trip_duration_minutes    0
dtype: int64
